File: android/src/main/kotlin/com/tartalabs/emotiondetection/sdk/EmotionFlutterBridgePayloads.kt
package com.tartalabs.emotiondetection.sdk

class EmotionFlutterBridgeException(message: String) : IllegalArgumentException(message)

data class EmotionRegisterModelPayload(
    val modelId: String,
    val resourceBase: String,
    val encryptedExtension: String = "onnx.enc",
    val hkdfInfo: String = "model_runtime",
    val masterKeyBase64: String? = null
) {
    companion object {
        fun from(arguments: Any?): EmotionRegisterModelPayload {
            val map = arguments.asMap()
            val modelId = map.requiredString("modelId")
            return EmotionRegisterModelPayload(
                modelId = modelId,
                resourceBase = map.stringOrNull("resourceBase") ?: modelId,
                encryptedExtension = map.stringOrNull("encExt") ?: "onnx.enc",
                hkdfInfo = map.stringOrNull("hkdfInfo") ?: "model_runtime",
                masterKeyBase64 = map.stringOrNull("masterKeyB64")
            )
        }
    }
}

data class EmotionPredictPayload(
    val modelId: String,
    val inputs: Map<String, Any?>
) {
    companion object {
        fun from(arguments: Any?): EmotionPredictPayload {
            val map = arguments.asMap()
            return EmotionPredictPayload(
                modelId = map.requiredString("modelId"),
                inputs = map.requiredMap("inputs")
            )
        }
    }
}

data class EmotionUserCodePayload(
    val userName: String?,
    val userCodeBase64: String,
    val modelId: String?,
    val requireBiometrics: Boolean = false
) {
    companion object {
        fun from(arguments: Any?): EmotionUserCodePayload {
            val map = arguments.asMap()
            return EmotionUserCodePayload(
                userName = map.stringOrNull("userName"),
                userCodeBase64 = map.requiredString("userCodeB64"),
                modelId = map.stringOrNull("modelId"),
                requireBiometrics = map["requireBiometrics"] as? Boolean ?: false
            )
        }
    }
}

private fun Any?.asMap(): Map<String, Any?> {
    @Suppress("UNCHECKED_CAST")
    return this as? Map<String, Any?> ?: throw EmotionFlutterBridgeException("arguments must be a map")
}

private fun Map<String, Any?>.requiredString(key: String): String {
    val value = this[key] as? String
    if (value.isNullOrBlank()) {
        throw EmotionFlutterBridgeException("$key is required")
    }
    return value
}

private fun Map<String, Any?>.stringOrNull(key: String): String? =
    (this[key] as? String)?.takeIf { it.isNotBlank() }

private fun Map<String, Any?>.requiredMap(key: String): Map<String, Any?> {
    @Suppress("UNCHECKED_CAST")
    return this[key] as? Map<String, Any?>
        ?: throw EmotionFlutterBridgeException("$key is required")
}

File: apple/EmotionNativeSDK/Sources/EmotionNativeSDK/EmotionFlutterBridgePayloads.swift
import Foundation

public enum EmotionFlutterBridgeError: Error, Equatable {
    case invalidArguments
    case missingString(String)
    case missingMap(String)
}

public struct EmotionRegisterModelPayload: Equatable {
    public var modelID: String
    public var resourceBase: String
    public var encryptedExtension: String
    public var hkdfInfo: String
    public var masterKeyBase64: String?

    public init(arguments: Any?) throws {
        guard let arguments = arguments as? [String: Any] else {
            throw EmotionFlutterBridgeError.invalidArguments
        }
        modelID = try Self.requiredString("modelId", in: arguments)
        resourceBase = try Self.requiredString("resourceBase", in: arguments)
        encryptedExtension = arguments["encExt"] as? String ?? "onnx.enc"
        hkdfInfo = arguments["hkdfInfo"] as? String ?? "model_runtime"
        masterKeyBase64 = arguments["masterKeyB64"] as? String
    }

    private static func requiredString(_ key: String, in arguments: [String: Any]) throws -> String {
        guard let value = arguments[key] as? String, !value.isEmpty else {
            throw EmotionFlutterBridgeError.missingString(key)
        }
        return value
    }
}

public struct EmotionPredictPayload: Equatable {
    public var modelID: String
    public var inputs: [String: Any]

    public init(arguments: Any?) throws {
        guard let arguments = arguments as? [String: Any] else {
            throw EmotionFlutterBridgeError.invalidArguments
        }
        guard let modelID = arguments["modelId"] as? String, !modelID.isEmpty else {
            throw EmotionFlutterBridgeError.missingString("modelId")
        }
        guard let inputs = arguments["inputs"] as? [String: Any] else {
            throw EmotionFlutterBridgeError.missingMap("inputs")
        }
        self.modelID = modelID
        self.inputs = inputs
    }

    public static func == (lhs: EmotionPredictPayload, rhs: EmotionPredictPayload) -> Bool {
        lhs.modelID == rhs.modelID && NSDictionary(dictionary: lhs.inputs).isEqual(to: rhs.inputs)
    }
}

public struct EmotionUserCodePayload: Equatable {
    public var userName: String
    public var userCodeBase64: String
    public var modelID: String?
    public var requireBiometrics: Bool

    public init(arguments: Any?) throws {
        guard let arguments = arguments as? [String: Any] else {
            throw EmotionFlutterBridgeError.invalidArguments
        }
        guard let userName = arguments["userName"] as? String, !userName.isEmpty else {
            throw EmotionFlutterBridgeError.missingString("userName")
        }
        guard let userCodeBase64 = arguments["userCodeB64"] as? String, !userCodeBase64.isEmpty else {
            throw EmotionFlutterBridgeError.missingString("userCodeB64")
        }
        self.userName = userName
        self.userCodeBase64 = userCodeBase64
        self.modelID = arguments["modelId"] as? String
        self.requireBiometrics = arguments["requireBiometrics"] as? Bool ?? false
    }
}
